File: projects/LeetcodeJavaSolutions/src/other_problems/checkIn03/Calculator.java
package other_problems.checkIn03;

import java.util.Stack;

public class Calculator {
    public int calculate(String s) {
        Stack<Integer> stack = new Stack<>();
        int num = 0;
        char opt = '+';
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (Character.isDigit(ch)) {
                num = num * 10 + (ch - '0');
            }
            if ((!Character.isDigit(ch) && ch != ' ') || i == s.length() - 1) {
                if (opt == '+') {
                    stack.push(num);
                } else if (opt == '-') {
                    stack.push(-num);
                } else if (opt == '*') {
                    stack.push(stack.pop() * num);
                } else {
                    stack.push(stack.pop() / num);
                }
                num = 0;
                opt = ch;
            }
        }
        int res = 0;
        while (!stack.isEmpty()) {
            res += stack.pop();
        }
        return res;
    }
    public static void main(String[] args) {
        String s = "14-3/2";
        // String s = "14/3*2";
        // String s = "42";
        // String s = " 3/2 ";
        // String s = "3+2*2";
        Calculator calc = new Calculator();
        int res = calc.calculate(s);
        System.out.println(res);
    }
}

File: LeetcodeJavaSolutions/src/other_problems/checkIn03/calculator.js
function isDigit(s) {
    return Number.isInteger(+s) && s !== ' '
}

function calculate(s) {
    const stack = [];
    let opt = '+';
    let num = 0;
    let index = 0;
    while (index < s.length) {
        const ch = s[index];
        if (isDigit(ch)) {
            num = num * 10 + (ch - 0);
        }
        if ((!isDigit(ch) && ch !== ' ') || index === s.length - 1) {
            if (opt === '+') {
                stack.push(num);
            } else if (opt === '-') {
                stack.push(-num);
            } else if (opt === '*') {
                stack.push(stack.pop() * num);
            } else if (opt === '/') {
                const last = stack.pop();
                let temp;
                if (last < 0) {
                    temp = -Math.floor(Math.abs(last) / num);
                } else {
                    temp = Math.floor(last / num);
                }
                stack.push(temp);
            }
            num = 0;
            opt = ch;
        }
        index++;
    }
    let res = 0;
    while (stack && stack.length > 0) {
        res += stack.pop();
    }
    return res;
}

function main() {
    // const s = "1*2-3/4+5*6-7*8+9/10";
    // const s = '14-3/2';
    // const s = '14/3*2';
    // const s = '42';
    // const s = ' 3/2 ';
    const s = '3+2*2';
    const ans = calculate(s);
    console.log('ans', ans)
}

main()